### ISSUE TITLE
libretro core slower than native PSP build

### ISSUE THREAD

**Issue by bmaupin** [libretro/picodrive#112]
Testing with Sonic 1, I get FPS dips as low as 51 with the libretro core with noticable audio popping/crackling.

With the native picodrive build I get near constant 60 FPS, not dipping below 59 FPS.

It does not seem to be caused by:

- Lack of frame skipping
    - I still get consistent framerate with the native Picodrive even with frame skipping turned off
- Vsync
    - Disabling vsync in RetroArch has no effect
- Which interpreter is used
    - The native build has the option of fast vs accurate interpreter; neither one exhibits framerate issues with Sonic 1
- CPU frequency
    - Both the libretro core and the native build seem to default to a CPU frequency of 333 MHz
- Sound rate
    - [The libretro core defaults to a sound rate of 44100](https://github.com/libretro/picodrive/blob/master/platform/libretro/libretro.c#L91); setting the native PSP build to 44100 has no noticeable impact on framerate
- Differences between upstream repo and fork
    > This branch is 258 commits ahead, 18 commits behind notaz:master.
    - I built the libretro core from notaz's repo and it exhibits the same problems

Needs investigation:
- Missing optimizations for PSP?
    - Does the native build have PSP-specific optimizations that are missing from the libretro core?
- Version differences?
    - I'm testing with the latest PSP native build on notaz's website (1.51b), which is > 10 years old. Maybe some of the changes to the core since then have negatively impacted performance?
- Other differences between core and native build?
    - Different defaults?
- ...what else?

**Comment by hizzlekizzle** [comment]
Can you build the libretro core from a very old commit (e.g., around when it was first libretro-ized)? This would at least tell you if it's changes to the core since that point in time.

**Comment by bmaupin** [comment]
I found a random native build of picodrive 1.92 for PSP online, did some testing, and this is what I found:

- Frame skipping was set to auto by default. I disabled this right away, which caused the audio to stutter much like I saw with this core.
- The audio quality was set to 44100Hz by default. Turning this down to 22050Hz did seem to improve the framerate and eliminate the audio stuttering again.
- Vsync (disabled by default) caused some tearing, but turning it on didn't cause any reduction in framerate
- The renderer is set to accurate by default. With the audio quality set to the default and frame skipping off, the accurate renderer had a slightly slower framerate enough to cause audio stuttering. Setting the renderer to fast increased the framerate enough to eliminate the audio stuttering.

Based on those tests, frame skipping, audio quality, and renderer all affected performance. I see frame skipping as a bit of a last resort. While the audio quality did improve performance, it seemed to use APIs specific to the PSP (and not included in the vanilla SDK if I recall correctly).

That leaves the renderer, so I went ahead and added an option to enable the fast renderer. I'll submit a PR soon.
